library/crypto/sha512.c

```diff
@@ -260,22 +260,28 @@ sha512_file (int fd, off_t offset, size_t count, void *resblock)
           if (sum == BLOCKSIZE)
             break;
 
-          if (n == 0) // eof
-              goto process_partial_block;
+          if (n == 0) { /* eof */
+              if (total + sum < count) {
+                  free (buffer);
+                  return -ESPIPE;
+              } else {
+                  goto process_partial_block;
+              }
+          }
         }
 
-      /* Process buffer with BLOCKSIZE bytes.  Note that
-                        BLOCKSIZE % 128 == 0
-       */
-      if (total + BLOCKSIZE < count) {
-          total += BLOCKSIZE;
+      if (total + sum < count) {
```

Comment:
sum = count - total;
  if (sum > 0)
    sha512_process_bytes (buffer, sum, &ctx);

sum is greater than buffer size, when EOF happens and offset+count > file_size.

I wrote this function and sha512_file_ctx() during mmap -> pread transition in eblob. And I missed this bug, because did not wrote tests.